src/Npgsql/PoolManager.cs

```diff
@@ -13,20 +15,49 @@ namespace Npgsql;
 /// </remarks>
 static class PoolManager
 {
-    internal static ConcurrentDictionary<string, NpgsqlDataSource> Pools { get; } = new();
+    static ConcurrentDictionary<string, NpgsqlDataSource> _pools = new();
+
+    // Keep the IEnumerable API for testing purposes but remove the rest to avoid careless (unlocked) access to the
+    // private dictionary.
+    internal static IEnumerable<KeyValuePair<string, NpgsqlDataSource>> Pools => _pools.ToImmutableDictionary();
 
     internal static void Clear(string connString)
     {
-        // TODO: Actually remove the pools from here, #3387 (but be careful of concurrency)
-        if (Pools.TryGetValue(connString, out var pool))
-            pool.Clear();
+        // ReSharper disable once InconsistentlySynchronizedField
+        var pools = _pools;
+        if (!pools.TryRemove(connString, out var pool))
+                return;
+
+        pool.Dispose();
+
+        // The same NpgsqlDataSource instance may be referenced via different connection strings
+        // in our dictionary, so we have to make sure that we remove all occurrences
+#if NET5_0_OR_GREATER
+        foreach (var pair in pools)
+            if (ReferenceEquals(pair.Value, pool))
+                pools.TryRemove(pair);
+#else
+        // Make sure nobody adds back a new pool with the same connection string
+        // as one we've already removed.
+        lock (pools)
+        {
+            foreach (var pair in pools)
+                if (ReferenceEquals(pair.Value, pool))
+                    pools.TryRemove(pair.Key, out _);
+        }
+#endif
     }
 
     internal static void ClearAll()
     {
-        // TODO: Actually remove the pools from here, #3387 (but be careful of concurrency)
-        foreach (var pool in Pools.Values)
-            pool.Clear();
+        ConcurrentDictionary<string, NpgsqlDataSource> pools;
+        lock (_pools)
+        {
+            pools = _pools;
+            _pools = new();
+        }
+        foreach (var pool in pools.Values)
+            pool.Dispose();
```

Comment:
There is a curious case with MultipleHosts. The way it's working right now is `NpgsqlMultiHostDataSource` has a list of `MultiHostDataSourceWrapper` which are actually referenced by connections. I think we shouldn't remove them from the pool here and instead do so in the `NpgsqlMultiHostDataSource.Dispose`.
Then there is also a question of calling `Clear` on `MultiHostDataSourceWrapper`...

Reply:
As it currently stands, only `NpgsqlDataSource` implements `Dispose()` and that one calls `Clear()` which is implemented by the derived classes.
`MultiHostDataSourceWrapper.Clear()` currently calls `Clear()` on the wrapped `NpgsqlMultiHostDataSource` which then calls `Clear()` on all the Per-Host-Pools.

I don't see a problem with disposing whatever DataSource we have in the `PoolManager`.
Am I missing something again?